rust/src/pgsql/logger.rs

```diff
@@ -0,0 +1,432 @@
+                let jb = log_response(&response)?;
+                js.append_object(&jb)?;
+            }
+            js.close()?;
+        }
+        else if !tx.responses.is_empty() {
+            let jb = log_response(&tx.responses[0])?;
+            js.set_object("response", &jb)?;
+        }
+    } else {
+        SCLogDebug!("Suricata logging a transaction without finding a request. Tx_id {:?}", &tx.tx_id);
+        if !tx.responses.is_empty() {
+            js.open_array("response")?;
+            for response in &tx.responses {
+                let jb = log_response(&response)?;
+                js.append_object(&jb)?;
+            }
+            js.close()?;
+        } else {
+            js.set_string("request", "request is empty")?;
+            js.set_string("response", "response is empty")?;
+            SCLogDebug!("Suricata created an empty PGSQL transaction");
+        }
+    }
+    Ok(())
+}
+
+fn log_request(req: &PgsqlFEMessage) -> Result<JsonBuilder, JsonError>
+{
+    let mut js = JsonBuilder::new_object();
+    js.open_object(&req.message_type_to_str())?;
+    match req {
+        PgsqlFEMessage::StartupMessage(
+            StartupPacket{
+                length: _,
+                proto_major,
+                proto_minor,
+                params}) =>
+        {
+            let proto = format!("{}.{}", proto_major, proto_minor);
+            js.set_string("protocol_version", &proto)?;
+            let jb = log_pgsql_parameters(params)?;
+            js.set_object("startup_parameters", &jb)?;
+        },
+        PgsqlFEMessage::SSLRequest(_) =>
+        {
+            js.set_string("message_type", "SSL Request")?;
+        },
+        PgsqlFEMessage::SASLInitialResponse(
+            SASLInitialResponsePacket{
+                identifier: _,
+                length: _,
+                auth_mechanism,
+                param_length: _,
+                sasl_param,
+            }) =>
+        {
+            js.set_string("authentication_mechanism", &auth_mechanism.to_string())?;
+            js.set_string_from_bytes("sasl_param", sasl_param)?;
+        },
+        PgsqlFEMessage::PasswordMessage(
+            RegularPacket{
+                identifier: _,
+                length: _,
+                payload,
+            }) |
+        PgsqlFEMessage::SASLResponse(
+            RegularPacket{
+                identifier: _,
+                length: _,
+                payload,
+            }) |
+        PgsqlFEMessage::SimpleQuery(
+            RegularPacket{
+                identifier: _,
+                length: _,
+                payload,
+            }) =>
+        {
+            js.set_string_from_bytes("payload", payload)?;
+        },
+        PgsqlFEMessage::Terminate(
+            TerminationMessage{
+                identifier: _,
+                length: _,
+            }) =>
+        {
+            js.set_string("message_type", "termination_token")?;
+        }
+    }
+    js.close()?;
+    js.close()?;
```

Comment:
The double close kind of looks odd.  But not always a sign of an issue, so just add comments as to what you are closing.